In [78]:
import seaborn as sns
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn import ensemble
from sklearn import datasets
from sklearn.utils import shuffle
from sklearn.metrics import mean_squared_error
from sklearn.ensemble import GradientBoostingRegressor
from __future__ import unicode_literals

import numbers

from scipy import stats
from scipy.sparse import csc_matrix
from scipy.sparse import csr_matrix
from scipy.sparse import issparse
from scipy.special import expit

from time import time



## Gradient Boosting Machines Learning


### Agenda:

* What is Gradient Boosting for regression?
* What is the model steps for training and prediction for scikit-learn?
* How this pattern be applied to other machine learning models?


#### Note:  GB builds an additive model in a forward stage-wise fashion; it allows for the optimization of arbitrary differentiable loss functions. In each stage a regression tree is fit on the negative gradient of the given loss function.

# Reviewing the Boston Housing Dataset

In [66]:
# Load data
bHousing = pd.read_csv('data/b_housing.csv')
bHousing.head()

Unnamed: 0     crim    zn  indus  chas    nox     rm   age     dis  rad  \
0           1  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1   
1           2  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2   
2           3  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2   
3           4  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3   
4           5  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3   

   tax  ptratio   black  lstat  medv  
0  296     15.3  396.90   4.98  24.0  
1  242     17.8  396.90   9.14  21.6  
2  242     17.8  392.83   4.03  34.7  
3  222     18.7  394.63   2.94  33.4  
4  222     18.7  396.90   5.33  36.2

In [67]:
bHousing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 15 columns):
Unnamed: 0    506 non-null int64
crim          506 non-null float64
zn            506 non-null float64
indus         506 non-null float64
chas          506 non-null int64
nox           506 non-null float64
rm            506 non-null float64
age           506 non-null float64
dis           506 non-null float64
rad           506 non-null int64
tax           506 non-null int64
ptratio       506 non-null float64
black         506 non-null float64
lstat         506 non-null float64
medv          506 non-null float64
dtypes: float64(11), int64(4)
memory usage: 59.4 KB


* 506 observation  
* 14 features, represented by the 14 column of data 
* Regression problem since all reponses are numeric


## Data Visualization

* data visualization is relevant in that it allow the users to explore and analyze data.

In [68]:
# view first 5 rows
bHousing.head()

Unnamed: 0     crim    zn  indus  chas    nox     rm   age     dis  rad  \
0           1  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1   
1           2  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2   
2           3  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2   
3           4  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3   
4           5  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3   

   tax  ptratio   black  lstat  medv  
0  296     15.3  396.90   4.98  24.0  
1  242     17.8  396.90   9.14  21.6  
2  242     17.8  392.83   4.03  34.7  
3  222     18.7  394.63   2.94  33.4  
4  222     18.7  396.90   5.33  36.2

In [69]:
# basic statistics
bHousing.describe()

Unnamed: 0        crim          zn       indus        chas         nox  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean   253.500000    3.613524   11.363636   11.136779    0.069170    0.554695   
std    146.213884    8.601545   23.322453    6.860353    0.253994    0.115878   
min      1.000000    0.006320    0.000000    0.460000    0.000000    0.385000   
25%    127.250000    0.082045    0.000000    5.190000    0.000000    0.449000   
50%    253.500000    0.256510    0.000000    9.690000    0.000000    0.538000   
75%    379.750000    3.677082   12.500000   18.100000    0.000000    0.624000   
max    506.000000   88.976200  100.000000   27.740000    1.000000    0.871000   

               rm         age         dis         rad         tax     ptratio  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     6.284634   68.574901    3.795043    9.549407  408.237154   18.455534   
std      0.702617   28.148861    2.105710    8.707259  168.537116    2.164946   
min      3.561000    2.900000    1.129600    1.000000  187.000000   12.600000   
25%      5.885500   45.025000    2.100175    4.000000  279.000000   17.400000   
50%      6.208500   77.500000    3.207450    5.000000  330.000000   19.050000   
75%      6.623500   94.075000    5.188425   24.000000  666.000000   20.200000   
max      8.780000  100.000000   12.126500   24.000000  711.000000   22.000000   

            black       lstat        medv  
count  506.000000  506.000000  506.000000  
mean   356.674032   12.653063   22.532806  
std     91.294864    7.141062    9.197104  
min      0.320000    1.730000    5.000000  
25%    375.377500    6.950000   17.025000  
50%    391.440000   11.360000   21.200000  
75%    396.225000   16.955000   25.000000  
max    396.900000   37.970000   50.000000

In [ ]:
sns.lmplot('tax', 'medv', bHousing)

In [74]:
# choosing variable for lmplot with medv

sns.lmplot('crim', 'medv', bHousing)

In [72]:
bHousing.corr()

Unnamed: 0      crim        zn     indus      chas       nox  \
Unnamed: 0    1.000000  0.407407 -0.103393  0.399439 -0.003759  0.398736   
crim          0.407407  1.000000 -0.200469  0.406583 -0.055892  0.420972   
zn           -0.103393 -0.200469  1.000000 -0.533828 -0.042697 -0.516604   
indus         0.399439  0.406583 -0.533828  1.000000  0.062938  0.763651   
chas         -0.003759 -0.055892 -0.042697  0.062938  1.000000  0.091203   
nox           0.398736  0.420972 -0.516604  0.763651  0.091203  1.000000   
rm           -0.079971 -0.219247  0.311991 -0.391676  0.091251 -0.302188   
age           0.203784  0.352734 -0.569537  0.644779  0.086518  0.731470   
dis          -0.302211 -0.379670  0.664408 -0.708027 -0.099176 -0.769230   
rad           0.686002  0.625505 -0.311948  0.595129 -0.007368  0.611441   
tax           0.666626  0.582764 -0.314563  0.720760 -0.035587  0.668023   
ptratio       0.291074  0.289946 -0.391679  0.383248 -0.121515  0.188933   
black        -0.295041 -0.385064  0.175520 -0.356977  0.048788 -0.380051   
lstat         0.258465  0.455621 -0.412995  0.603800 -0.053929  0.590879   
medv         -0.226604 -0.388305  0.360445 -0.483725  0.175260 -0.427321   

                  rm       age       dis       rad       tax   ptratio  \
Unnamed: 0 -0.079971  0.203784 -0.302211  0.686002  0.666626  0.291074   
crim       -0.219247  0.352734 -0.379670  0.625505  0.582764  0.289946   
zn          0.311991 -0.569537  0.664408 -0.311948 -0.314563 -0.391679   
indus      -0.391676  0.644779 -0.708027  0.595129  0.720760  0.383248   
chas        0.091251  0.086518 -0.099176 -0.007368 -0.035587 -0.121515   
nox        -0.302188  0.731470 -0.769230  0.611441  0.668023  0.188933   
rm          1.000000 -0.240265  0.205246 -0.209847 -0.292048 -0.355501   
age        -0.240265  1.000000 -0.747881  0.456022  0.506456  0.261515   
dis         0.205246 -0.747881  1.000000 -0.494588 -0.534432 -0.232471   
rad        -0.209847  0.456022 -0.494588  1.000000  0.910228  0.464741   
tax        -0.292048  0.506456 -0.534432  0.910228  1.000000  0.460853   
ptratio    -0.355501  0.261515 -0.232471  0.464741  0.460853  1.000000   
black       0.128069 -0.273534  0.291512 -0.444413 -0.441808 -0.177383   
lstat      -0.613808  0.602339 -0.496996  0.488676  0.543993  0.374044   
medv        0.695360 -0.376955  0.249929 -0.381626 -0.468536 -0.507787   

               black     lstat      medv  
Unnamed: 0 -0.295041  0.258465 -0.226604  
crim       -0.385064  0.455621 -0.388305  
zn          0.175520 -0.412995  0.360445  
indus      -0.356977  0.603800 -0.483725  
chas        0.048788 -0.053929  0.175260  
nox        -0.380051  0.590879 -0.427321  
rm          0.128069 -0.613808  0.695360  
age        -0.273534  0.602339 -0.376955  
dis         0.291512 -0.496996  0.249929  
rad        -0.444413  0.488676 -0.381626  
tax        -0.441808  0.543993 -0.468536  
ptratio    -0.177383  0.374044 -0.507787  
black       1.000000 -0.366087  0.333461  
lstat      -0.366087  1.000000 -0.737663  
medv        0.333461 -0.737663  1.000000

In [79]:
# correlation heatmap

#plt.figure(bHousing.corr(), annot=True)
for train_index, test_index in bHousing:
    # do something with slices
    df.loc[train_index]
    df.loc[test_index]

ValueError: too many values to unpack